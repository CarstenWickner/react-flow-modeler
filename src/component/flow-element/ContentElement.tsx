import * as React from "react";

export const ContentElement: React.FunctionComponent<{
    children: React.ReactChild;
}> = ({ children }) => (
    <>
        <div className="arrow" />
        <div className="flow-element content-element">{children}</div>
    </>
);
