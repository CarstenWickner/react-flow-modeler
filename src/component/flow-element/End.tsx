import * as React from "react";

export const End: React.FunctionComponent = () => (
    <>
        <div className="arrow" />
        <div className="flow-element end-element" />
    </>
);
