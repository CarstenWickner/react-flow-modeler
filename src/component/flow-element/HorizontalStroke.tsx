import * as React from "react";

import { ConnectionType } from "../../types/GridCellData";

const getConnectionClassName = (connectionType: ConnectionType): string => {
    switch (connectionType) {
        case ConnectionType.First:
            return "bottom-half";
        case ConnectionType.Middle:
            return "full-height";
        case ConnectionType.Last:
            return "top-half";
    }
};

export class HorizontalStroke extends React.Component<
    { className?: string; children?: React.ReactNode } & (
        | { incomingConnection: ConnectionType; outgoingConnection?: never }
        | { incomingConnection?: never; outgoingConnection?: never }
        | { incomingConnection?: never; outgoingConnection: ConnectionType }
    ),
    { wrapperTopHeight: number }
> {
    readonly topLabelRef = React.createRef<HTMLDivElement>();
    state = { wrapperTopHeight: 0 };

    componentDidMount(): void {
        // trigger consideration of element height already after initial rendering
        this.componentDidUpdate();
    }

    componentDidUpdate(): void {
        if (this.topLabelRef.current && this.state.wrapperTopHeight !== this.topLabelRef.current.clientHeight) {
            // setting to state mostly to re-trigger rendering (the value could also be accessed directly during render())
            this.setState({ wrapperTopHeight: this.topLabelRef.current.clientHeight });
        }
    }

    render(): React.ReactNode {
        const { className, incomingConnection, outgoingConnection, children } = this.props;
        const classNameSuffix = className ? ` ${className}` : "";
        return (
            <>
                {incomingConnection && <div className={`stroke-vertical ${getConnectionClassName(incomingConnection)}${classNameSuffix}`} />}
                {!children && <div className={`stroke-horizontal${classNameSuffix}`} />}
                {children && (
                    <div className={`centered-line-wrapper${classNameSuffix}`}>
                        <div className="wrapper-top-label" ref={this.topLabelRef}>
                            {children}
                        </div>
                        <div className={`stroke-horizontal${classNameSuffix}`} />
                        <div
                            className="wrapper-bottom-spacing"
                            style={(this.topLabelRef.current && { height: `${this.state.wrapperTopHeight}px` }) || undefined}
                        />
                    </div>
                )}
                {outgoingConnection && <div className={`stroke-vertical ${getConnectionClassName(outgoingConnection)}${classNameSuffix}`} />}
            </>
        );
    }
}
