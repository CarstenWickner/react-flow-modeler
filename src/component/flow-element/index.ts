export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { Gateway } from "./Gateway";
export { GatewayConnector } from "./GatewayConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
