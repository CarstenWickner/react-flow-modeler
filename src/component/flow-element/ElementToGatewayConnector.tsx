import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";
import { ConnectionType } from "../../types/GridCellData";

export const ElementToGatewayConnector: React.FunctionComponent<{
    connectionType: ConnectionType;
}> = ({ connectionType }) => <HorizontalStroke outgoingConnection={connectionType} />;
