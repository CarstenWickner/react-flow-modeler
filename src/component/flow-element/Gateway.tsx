import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export const Gateway: React.FunctionComponent<{ type: "converging" } | { type: "diverging"; children?: React.ReactChild }> = ({ type, children }) => (
    <>
        <div className="arrow" />
        <div className={`flow-element gateway-element ${type}`} />
        {type === "diverging" && <HorizontalStroke>{children}</HorizontalStroke>}
    </>
);
