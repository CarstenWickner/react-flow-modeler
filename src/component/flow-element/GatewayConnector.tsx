import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export const GatewayConnector: React.FunctionComponent<{
    incomingConnection: "first" | "middle" | "last";
    children?: React.ReactChild;
}> = ({ incomingConnection, children }) => <HorizontalStroke incomingConnection={incomingConnection}>{children}</HorizontalStroke>;
