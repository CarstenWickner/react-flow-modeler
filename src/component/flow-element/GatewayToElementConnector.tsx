import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";
import { ConnectionType } from "../../types/GridCellData";

export const GatewayToElementConnector: React.FunctionComponent<{
    connectionType: ConnectionType;
    children?: React.ReactChild;
}> = ({ connectionType, children }) => <HorizontalStroke incomingConnection={connectionType}>{children}</HorizontalStroke>;
