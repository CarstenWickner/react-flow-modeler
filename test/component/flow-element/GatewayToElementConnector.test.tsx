import * as React from "react";
import { shallow } from "enzyme";

import { GatewayToElementConnector } from "../../../src/component/flow-element/GatewayToElementConnector";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const component = shallow(<GatewayToElementConnector connectionType="first">label</GatewayToElementConnector>);
        expect(component).toMatchSnapshot();
    });
});
