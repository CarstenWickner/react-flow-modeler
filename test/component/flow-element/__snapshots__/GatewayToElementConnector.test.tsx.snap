// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<HorizontalStroke
  incomingConnection={1}
  outgoingConnection="single"
>
  label
</HorizontalStroke>
`;
